runtime/moonbeam/src/timestamp.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A way to get a relyable timestamp
+
+use crate::Runtime;
+use cumulus_pallet_parachain_system::{
+	consensus_hook::UnincludedSegmentCapacity,
+	relay_state_snapshot::{self, ReadEntryErr},
+	ConsensusHook, RelayChainStateProof,
+};
+use frame_support::pallet_prelude::*;
+use frame_support::storage::types::{StorageValue, ValueQuery};
+use frame_support::traits::{StorageInstance, Time};
+pub use moonbeam_core_primitives::well_known_relay_keys;
+
+/// Get the relay timestamp.
+/// Noe that the relay timestamp is populated at the parachain system inherent.
+/// If you fetch the timestamp before, you will get the timestamp of the parent block.
+pub struct RelayTimestamp;
+impl Time for RelayTimestamp {
+	type Moment = u64;
+
+	fn now() -> Self::Moment {
+		RelayTimestampNow::get()
+	}
+}
+
+/// A wrapper around the consensus hook to get the relay timlestmap from the relay storage proof
+pub struct ConsensusHookWrapperForRelayTimestamp<Inner>(core::marker::PhantomData<Inner>);
```

Comment:
Note: this file will be removed from this PR once https://github.com/moonbeam-foundation/moonbeam/pull/2804 is merged (as we will use the `ConsensusHookWrapperForRelayTimestamp` struct present in `runtime/common/src/timestamp.rs`)